From 5577f21f07843d7457be79afb530e5739fbec465 Mon Sep 17 00:00:00 2001
From: ChenZiyu <czy199896@163.com>
Date: Fri, 10 Apr 2026 17:33:11 +0800
Subject: [PATCH 4/5] egl: fix sw fallback rejection in non-sw
 EGL_PLATFORM=device

previously progress could still be made during sw fallback here,
which would lead to unpredictable results with driver loading e.g., crashing

cc: mesa-stable
Part-of: <mesa/mesa!34609>
Origin: upstream, commit 8a339cdebccdea0610bdd7a1ecc9a5ec63951940

Signed-off-by: ChenZiyu <czy199896@163.com>
---
 src/egl/drivers/dri2/platform_device.c | 25 +++++++++++++++----------
 1 file changed, 15 insertions(+), 10 deletions(-)

diff --git a/src/egl/drivers/dri2/platform_device.c b/src/egl/drivers/dri2/platform_device.c
index 41ab931e..5fb3fadf 100644
--- a/src/egl/drivers/dri2/platform_device.c
+++ b/src/egl/drivers/dri2/platform_device.c
@@ -284,16 +284,21 @@ device_probe_device(_EGLDisplay *disp)
    if (!dri2_dpy->driver_name)
       goto err_name;
 
-   /* When doing software rendering, some times user still want to explicitly
-    * choose the render node device since cross node import doesn't work between
-    * vgem/virtio_gpu yet. It would be nice to have a new EXTENSION for this.
-    * For now, just fallback to kms_swrast. */
-   if (disp->Options.ForceSoftware && !request_software &&
-       (strcmp(dri2_dpy->driver_name, "vgem") == 0 ||
-        strcmp(dri2_dpy->driver_name, "virtio_gpu") == 0)) {
-      free(dri2_dpy->driver_name);
-      _eglLog(_EGL_WARNING, "NEEDS EXTENSION: falling back to kms_swrast");
-      dri2_dpy->driver_name = strdup("kms_swrast");
+   /* this is software fallback */
+   if (disp->Options.ForceSoftware && !request_software) {
+      /* When doing software rendering, some times user still want to explicitly
+      * choose the render node device since cross node import doesn't work between
+      * vgem/virtio_gpu yet. It would be nice to have a new EXTENSION for this.
+      * For now, just fallback to kms_swrast. */
+      if (strcmp(dri2_dpy->driver_name, "vgem") == 0 ||
+          strcmp(dri2_dpy->driver_name, "virtio_gpu") == 0) {
+         free(dri2_dpy->driver_name);
+         _eglLog(_EGL_WARNING, "NEEDS EXTENSION: falling back to kms_swrast");
+         dri2_dpy->driver_name = strdup("kms_swrast");
+      } else if (strcmp(dri2_dpy->driver_name, "vmwgfx")) {
+         /* this is software fallback; deny progress since a hardware device was requested */
+         return false;
+      }
    }
 
    if (!dri2_load_driver(disp))
-- 
2.20.1

